domain/learningandworkprogress/src/main/kotlin/uk/gov/justice/digital/hmpps/domain/learningandworkprogress/education/PreviousQualifications.kt

```diff
@@ -24,20 +26,15 @@ data class PreviousQualifications(
   val lastUpdatedAtPrison: String,
 )
 
-enum class EducationLevel {
-  PRIMARY_SCHOOL,
-  SECONDARY_SCHOOL_LEFT_BEFORE_TAKING_EXAMS,
-  SECONDARY_SCHOOL_TOOK_EXAMS,
-  FURTHER_EDUCATION_COLLEGE,
-  UNDERGRADUATE_DEGREE_AT_UNIVERSITY,
-  POSTGRADUATE_DEGREE_AT_UNIVERSITY,
-  NOT_SURE,
-}
-
 data class Qualification(
+  val reference: UUID?,
   val subject: String,
   val level: QualificationLevel,
   val grade: String,
+  val createdBy: String?,
```

Comment:
Obviously I don't know your code base so this may be standard throughout and a genuine reason.

Why would createdBy and createdAt ever be null?

Reply:
Good challenge 👍 
Yes, they could both be null.
We have a concept of a "domain" in our codebase, which sits between the REST API model classes, and the JPA entity classes.
When a new Qualification (or any other domain class) is created, it is not created until it has been persisted to the database; and we use the Spring annotations on the fields on the JPA entity classes to set the createdBy and createdAt values.

So, when a new Qualification is created, it comes in as a request in the REST API. That request object contains the data associated with a Qualification (grade, level, subject), but does not contain the DPS user or created time (ie. we dont expect the UI client to tell the API when something was created or by who)
The REST API request model object is mapped to this domain object (and therefore wont have createdBy and createdAt values, hence them being nullable)
The domain object is then mapped to the JPA entity which using the Spring annotations will set createdBy and createdAt

On the way back out, eg a REST API GET request, the JPA entity is mapped to the domain class, and this time it will have createdBy and createdAt populated. The domain class is then mapped to the REST API response model class where createdBy and createdAt are not nullable 👍